File: apps/dashboard/src/types/api/system.ts
import { z } from 'zod';

export const ComponentStatusSchema = z.object({
  name: z.string(),
  status: z.enum(['ok', 'degraded', 'unavailable']),
  state: z.string().nullable().optional(),
  latency_ms: z.number().nonnegative().nullable().optional(),
  updated_at: z.string().datetime().nullable().optional(),
  uptime_seconds: z.number().nonnegative().nullable().optional(),
  detail: z.string().nullable().optional(),
});

export type ComponentStatus = z.infer<typeof ComponentStatusSchema>;

export const SystemHealthSchema = z.object({
  status: z.enum(['ok', 'degraded', 'unavailable']),
  api: ComponentStatusSchema,
  database: ComponentStatusSchema,
  delta_rest: ComponentStatusSchema,
  delta_ws: ComponentStatusSchema,
  event_bus: ComponentStatusSchema,
  state_manager: ComponentStatusSchema,
});

export type SystemHealth = z.infer<typeof SystemHealthSchema>;

export const DeltaConnectionStateSchema = z.object({
  state: z.enum(['stopped', 'connecting', 'connected', 'disconnected']),
  connected: z.boolean(),
  authenticated: z.boolean(),
  public: z.boolean(),
  subscriptions: z.array(z.string()),
  requested_subscriptions: z.array(z.string()),
  last_message_at: z.string().datetime().nullable(),
  last_heartbeat_at: z.string().datetime().nullable(),
  connected_at: z.string().datetime().nullable(),
  messages_received: z.number().int().nonnegative(),
  connection_attempts: z.number().int().nonnegative(),
  reconnects: z.number().int().nonnegative(),
  uptime_seconds: z.number().nonnegative().nullable(),
});

export type DeltaConnectionState = z.infer<typeof DeltaConnectionStateSchema>;

export const SystemStatusSchema = z.object({
  status: z.enum(['ok', 'degraded', 'unavailable']),
  started_at: z.string().datetime(),
  uptime_seconds: z.number().nonnegative(),
  version: z.string(),
  environment: z.string(),
  market_data_live: z.boolean(),
  delta_ws_connected: z.boolean(),
  delta_ws: DeltaConnectionStateSchema.nullable(),
  last_ws_message_at: z.string().datetime().nullable(),
  last_heartbeat_at: z.string().datetime().nullable(),
  last_ws_reconnect_at: z.string().datetime().nullable(),
  last_rest_request_at: z.string().datetime().nullable(),
  last_ingestion_at: z.string().datetime().nullable(),
  symbols_tracked: z.number().int().nonnegative(),
});

export type SystemStatus = z.infer<typeof SystemStatusSchema>;

export const SystemMetricsSchema = z.object({
  collected_at: z.string().datetime(),
  synchronized_markets: z.number().int().nonnegative().nullable(),
  stored_candles: z.number().int().nonnegative().nullable(),
  messages_received: z.number().int().nonnegative(),
  messages_normalized: z.number().int().nonnegative(),
  validation_failures: z.number().int().nonnegative(),
  unsupported_messages: z.number().int().nonnegative(),
  events_published: z.number().int().nonnegative(),
  average_pipeline_latency_ms: z.number().nonnegative().nullable(),
  state_updates: z.number().int().nonnegative(),
  invalid_events: z.number().int().nonnegative(),
  state_symbols_tracked: z.number().int().nonnegative(),
  cache_hits: z.number().int().nonnegative(),
  cache_misses: z.number().int().nonnegative(),
  state_latest_update_at: z.string().datetime().nullable(),
  state_average_update_latency_ms: z.number().nonnegative().nullable(),
  state_order_books_cached: z.number().int().nonnegative(),
  state_trades_cached: z.number().int().nonnegative(),
  state_tickers_cached: z.number().int().nonnegative(),
  state_candles_cached: z.number().int().nonnegative(),
  state_latest_prices: z.record(z.string(), z.string()),
  event_bus_pending: z.number().int().nonnegative(),
  event_bus_subscribers: z.number().int().nonnegative(),
  event_bus_published: z.number().int().nonnegative(),
  event_bus_failed_handlers: z.number().int().nonnegative(),
  event_bus_average_handler_latency_ms: z.number().nonnegative().nullable(),
});

export type SystemMetrics = z.infer<typeof SystemMetricsSchema>;

export type SystemStatusValue = 'ok' | 'degraded' | 'unavailable';

File: services/api/app/schemas/system.py
"""Response schemas for the platform system endpoints.

Health reports per-component statuses; status reports process and data
freshness; metrics reports counters and aggregates. DB-derived fields are
``None`` when no database is configured, never fabricated.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

ComponentStatusValue = Literal["ok", "degraded", "unavailable"]


class ComponentStatus(BaseModel):
    """Status of a single platform component.

    ``state``, ``latency_ms``, ``updated_at``, and ``uptime_seconds`` are
    optional richer signals the dashboard renders when present.
    """

    name: str
    status: ComponentStatusValue
    state: str | None = None
    latency_ms: float | None = None
    updated_at: datetime | None = None
    uptime_seconds: float | None = None
    detail: str | None = None


class DeltaConnectionState(BaseModel):
    """Live WebSocket connection state reported from the runtime."""

    state: Literal["stopped", "connecting", "connected", "disconnected"]
    connected: bool
    authenticated: bool
    public: bool
    subscriptions: list[str] = []
    requested_subscriptions: list[str] = []
    last_message_at: datetime | None = None
    last_heartbeat_at: datetime | None = None
    connected_at: datetime | None = None
    messages_received: int = 0
    connection_attempts: int = 0
    reconnects: int = 0
    uptime_seconds: float | None = None


class SystemHealthResponse(BaseModel):
    """Per-component health of the platform."""

    status: ComponentStatusValue
    api: ComponentStatus
    database: ComponentStatus
    delta_rest: ComponentStatus
    delta_ws: ComponentStatus
    event_bus: ComponentStatus
    state_manager: ComponentStatus


class SystemStatusResponse(BaseModel):
    """Process and data freshness overview."""

    status: ComponentStatusValue
    started_at: datetime
    uptime_seconds: float
    version: str
    environment: str
    market_data_live: bool
    delta_ws_connected: bool
    delta_ws: DeltaConnectionState | None = None
    last_ws_message_at: datetime | None = None
    last_heartbeat_at: datetime | None = None
    last_ws_reconnect_at: datetime | None = None
    last_rest_request_at: datetime | None = None
    last_ingestion_at: datetime | None = None
    symbols_tracked: int


class SystemMetricsResponse(BaseModel):
    """Counters and aggregates for the platform stack."""

    collected_at: datetime
    synchronized_markets: int | None = None
    stored_candles: int | None = None
    messages_received: int = 0
    messages_normalized: int = 0
    validation_failures: int = 0
    unsupported_messages: int = 0
    events_published: int = 0
    average_pipeline_latency_ms: float | None = None
    state_updates: int = 0
    invalid_events: int = 0
    state_symbols_tracked: int = 0
    cache_hits: int = 0
    cache_misses: int = 0
    state_latest_update_at: datetime | None = None
    state_average_update_latency_ms: float | None = None
    state_order_books_cached: int = 0
    state_trades_cached: int = 0
    state_tickers_cached: int = 0
    state_candles_cached: int = 0
    state_latest_prices: dict[str, str] = {}
    event_bus_pending: int = 0
    event_bus_subscribers: int = 0
    event_bus_published: int = 0
    event_bus_failed_handlers: int = 0
    event_bus_average_handler_latency_ms: float | None = None
